libs/geometry/src/TangentSpaceMesh.cpp

```diff
@@ -187,14 +175,155 @@ Algorithm selectAlgorithm(TangentSpaceMeshInput *input) {
+    if (abs(N.x) > abs(N.z) + std::numeric_limits<float>::epsilon()) {
+        T = float3{-N.y, N.x, 0.0f};
+    } else {
+        T = float3{0.0f, -N.z, N.y};
+    }
+    T = normalize(T);
+    B = cross(N, T);
+}
+
+void normalsOnlyMethod(const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output,
+    NormalsOnlyKernelPtr kernel) noexcept {
+    const size_t vertexCount = input->vertexCount;
+    quatf* quats = new quatf[vertexCount];
+
+    const float3* normal = input->normals;
+    size_t nstride = input->normalStride ? input->normalStride : sizeof(float3);
+
+    for (size_t qindex = 0; qindex < vertexCount; ++qindex) {
+        float3 n = *normal;
+        float3 b, t;
+        kernel(n, t, b);
+        quats[qindex] = mat3f::packTangentFrame({t, b, n});
+        normal = (const float3*) (((const uint8_t*) normal) + nstride);
+    }
+    output->tangentSpace = quats;
+    output->vertexCount = input->vertexCount;
+    output->triangleCount = input->triangleCount;
+    output->uvs = input->uvs;
+    output->positions = input->positions;
+    output->triangles32 = input->triangles32;
+    output->triangles16 = input->triangles16;
+}
+
+template<typename CastType, typename InputType>
+inline CastType pointerAdd(const InputType* ptr, size_t index, size_t stride) {
+    return (*((CastType*) (((const uint8_t*) ptr) + index * stride)));
+}
+
+void flatShadingMethod(const TangentSpaceMeshInput* input, TangentSpaceMeshOutput* output)
+        noexcept {
+    const float3* positions = input->positions;
+    const size_t pstride = input->positionStride ? input->positionStride : sizeof(float3);
+
+    const float2* uvs = input->uvs;
+    const size_t uvstride = input->uvStride ? input->uvStride : sizeof(float2);
+
+    const bool isTriangle16 = input->triangles16 != nullptr;
+    const uint8_t* triangles = isTriangle16 ? (const uint8_t*) input->triangles16 :
+            (const uint8_t*) input->triangles32;
+
+    const size_t triangleCount = input->triangleCount;
+    const size_t tstride = isTriangle16 ? sizeof(ushort3) : sizeof(uint3);
+
+    const size_t outVertexCount = triangleCount * 3;
+    float3* outPositions = new float3[outVertexCount];
+    float2* outUvs = uvs ? new float2[outVertexCount] : nullptr;
+    quatf* quats = new quatf[outVertexCount];
+
+    const size_t outTriangleCount = triangleCount;
+    uint3* outTriangles = new uint3[outTriangleCount];
+
+    size_t vindex = 0;
+    for (size_t tindex = 0; tindex < triangleCount; ++tindex) {
+        uint3 tri = isTriangle16 ?
+                uint3(pointerAdd<ushort3, uint8_t>(triangles, tindex, tstride)) :
+                pointerAdd<uint3, uint8_t>(triangles, tindex, tstride);
+
+        const float3 pa = pointerAdd<float3, float3>(positions, tri.x, pstride);
+        const float3 pb = pointerAdd<float3, float3>(positions, tri.y, pstride);
+        const float3 pc = pointerAdd<float3, float3>(positions, tri.z, pstride);
```

Comment:
if you change `pointerAdd` so it takes only one template parameter then you won't need to specify it here, because it'll be deduced.

Reply:
done